### DISCUSSION TITLE
Fluent API support

### DISCUSSION THREAD

**Discussion by mattleibow** [CommunityToolkit/Maui.Markup#169] [New Feature Discussions]
---
**Issue moved from dotnet/maui#12380**
- Please respond to @idexus.

---

_From @idexus on Monday, January 2, 2023 8:09:40 PM_

### Discussed in https://github.com/dotnet/maui/discussions/12273

# Description

I would like to have fluent API support in MAUI so first I created a library [https://github.com/idexus/Sharp.UI](https://github.com/idexus/Sharp.UI) then because some classes are sealed, I decided to create a fork [https://github.com/idexus/maui](https://github.com/idexus/maui) to fully support this feature in the MAUI project. Most code is generated using source generators.

I have two questions
- Is it possible to include fluent API support directly in the maui project?
- If there is no way to add such support to the maui project for now, is there a way to "unseal" the maui classes?

I think for many it would make it much easier to create an interface without the need for XAML, without disabling the possibility of using it.

My goal was:

```cs
new ScrollView
{
    new VerticalStackLayout
    {
        new Grid
        {
            new Label("Hello")
                .Column(0)
                .FontSize(28),
            
            new Label("World", out var label)
                .Column(1),

            new Button("Click me")
                .Row(1)
                .ColumnSpan(2)
                .OnClicked(button => {
                    label.Text = "you";
                });
        }
        .ColumnDefinitions(e => e.Star(2).Star(1))
        .RowDefinitions(e => e.Star().Absolute(100)),
            
        new Path
        {
            new GeometryGroup
            {
                new PathGeometry
                {
                    new PathFigure(15, 50)
                    {
                        new LineSegment(800,150),
                        new LineSegment(500,50)
                    }
                },

                new EllipseGeometry()
                    .Center(new Point(50,70))
                    .RadiusX(10)
                    .RadiusY(50),
            }
        }
        .Stroke(Colors.Yellow)
        .Fill(Colors.Red),
    }
}
```

Unfortunately some classes are sealed, e.g. 

`PathGeometry`, `PathFigure`, `TapGestureRecognizer`, `PinchGestureRecognizer`, `PointerGestureRecognizer`, `SwipeGestureRecognizer`, `Ellipse`, `Polyline`, `Line`, `Path`, `Polygon`, `Rectangle`, `RoundRectangle`, `TableSection`, `ColumnDefinition`, `RowDefinition` `Style`, `Trigger`,... etc.

If I want to create a library with these classes so that I could create the UI declaratively as I described, only in code using fluent interface, I can't do it without wrapping them.

# Detailed description

Below is a description of how each problem was solved.

## Properties

Fluent extension methods are generated for all properties and bindable properties (for all derived classes of `BindableObject`, and for `Style`, `VisualState`, `VisualStateGroup`, `VisualStateGroupList` classes)

Usage:
```cs
Label()
  .FontSize(28)
  .TextColor(Colors.White)
```
Generated methods for `FontSize` property:
```cs
public static T FontSize<T>(this T obj,
    double fontSize)
    where T : Microsoft.Maui.Controls.Label
{
    obj.FontSize = fontSize;
    return obj;
}

public static T FontSize<T>(this T obj,
    System.Func<ValueBuilder<double>, ValueBuilder<double>> buidValue)
    where T : Microsoft.Maui.Controls.Label
{
    var builder = buidValue(new ValueBuilder<double>());
    if (builder.ValueIsSet()) obj.FontSize = builder.GetValue();
    return obj;
}

public static T FontSize<T>(this T obj,
    System.Func<LazyValueBuilder<double>, LazyValueBuilder<double>> buidValue)
    where T : Microsoft.Maui.Controls.Label
{
    var builder = buidValue(new LazyValueBuilder<double>());
    if (builder.ValueIsSet()) obj.FontSize = builder.GetValue();
    return obj;
}

public static T FontSize<T>(this T obj,
    System.Func<BindingBuilder<double>, BindingBuilder<double>> buidBinding)
    where T : Microsoft.Maui.Controls.Label
{
    var builder = buidBinding(new BindingBuilder<double>(obj, Microsoft.Maui.Controls.Label.FontSizeProperty));
    builder.BindProperty();
    return obj;
}
```

## `EventHandler`

Fluent extension methods are generated for each `EventHandler` adding an `On` prefix

usage example:

```cs
new VerticalStackLayout
{
    new Label("Hello"),
    new Button("Click me")
        OnClicked(button => {

        });
}
```

generated methods for `Clicked` event handler:

```cs
public static T OnClicked<T>(this T obj, System.EventHandler handler)
    where T : Microsoft.Maui.Controls.Button
{
    obj.Clicked += handler;
    return obj;
}

public static T OnClicked<T>(this T obj, System.Action<T> action)
    where T : Microsoft.Maui.Controls.Button
{
    obj.Clicked += (o, arg) => action(obj);
    return obj;
}
```

## `ContentProperty` attributes

Implementation of `IList` (if not implemented yet) or `IEnumerable` interfaces to add `Add()` method support to class are generated for all classes with the `ContentProperty` attribute.

Example usage:

```cs
new Border
{
    new VerticalStackLayout
    {
        new Label("Hello").FontSize(28),
        new Grid
        {
            ...
        },
        new HorizontalStackLayout
        {
            new Label("foo"),
            new Button("OK").OnClicked(OnClickedButton)
        }

    }
}
```

Generated class implementation for single item containers:

```cs
public partial class Border : IEnumerable
{

    ...

    // ----- single item container -----

    public IEnumerator GetEnumerator() { yield return this.Content; }

    public void Add(Microsoft.Maui.Controls.View? content) => this.Content = content;

}
```

Generated class implementation for collection containers:

```cs
public partial class Shell : IList<Microsoft.Maui.Controls.ShellItem>
{

    ...

    // ----- collection container -----

    public int Count => this.Items.Count;
    public Microsoft.Maui.Controls.ShellItem this[int index] { get => this.Items[index]; set => this.Items[index] = value; }
    public bool IsReadOnly => false;
    public void Add(Microsoft.Maui.Controls.ShellItem item) => this.Items.Add(item);
    public void Clear() => this.Items.Clear();
    public bool Contains(Microsoft.Maui.Controls.ShellItem item) => this.Items.Contains(item);
    public void CopyTo(Microsoft.Maui.Controls.ShellItem[] array, int arrayIndex) => this.Items.CopyTo(array, arrayIndex);
    public IEnumerator<Microsoft.Maui.Controls.ShellItem> GetEnumerator() => this.Items.GetEnumerator();
    public int IndexOf(Microsoft.Maui.Controls.ShellItem item) => this.Items.IndexOf(item);
    public void Insert(int index, Microsoft.Maui.Controls.ShellItem item) => this.Items.Insert(index, item);
    public bool Remove(Microsoft.Maui.Controls.ShellItem item) => this.Items.Remove(item);
    public void RemoveAt(int index) => this.Items.RemoveAt(index);
    IEnumerator IEnumerable.GetEnumerator() => this.Items.GetEnumerator();

}
```

### Aditional constructors

Additional constructors have been added to simplify the interface creation process

example usage:

```cs
new VerticalStackLayout
{
    new Label("Hello"),
    new Button("Click me")
        OnClicked(button => {

        });
}
```

implementation

```cs
public partial class Label
{
    public Label(string text) : this()
    {
        this.Text = text;
    }
}
```

### Additional `out` parameter

Additional constructors with the `out` parameter are generated for all public constructors.

usage:

```cs
new VerticalStackLayout
{
    new Label("Hello", out var label)
    new Button("Click me")
        OnClicked(button => {
            label.Text = "Clicked";
        });
}
```
genetated constructor:

```cs
public partial class Label : IEnumerable
{

    // ----- constructors -----

    public Label(out Label label) : this()
    {
        label = this;
    }

    public Label(string text, out Label label) : this(text)
    {
        label = this;
    }

    ...
}
```

### `BindingBuilder`

For in-line creation of bindings

usage:

```cs
new VerticalStackLayout
{
    new Slider(1, 100, 1, out var slider),
    new Label()
        .Text(e => e.Path("Value").Source(slider).StringFormat("Value {0:F1}"))
}
```

### `ValueBuilder`

To set values depending on the app theme, platform or device idiom

```cs
new Label("Hello")
    .TextColor(e => e.OnDark(Colors.White).OnLight(Colors.Black))
    .FontSize(e => e.Default(28).OnPhone(18).OnTV(40))
```

### `LazyValueBuilder`

To lazily set values depending on the app theme, platform or device idiom

```cs
new Label()
    .Text(e => e
        .Default(() => DefaultName())
        .OnAndroid(() => AndroidName())
    )
```

### `ColumnDefinitionBuilder` and `RowDefinitionBuilder`

To define the number and sizes of rows and columns.

example:

```cs
new Grid
{
    ...
}
.RowDefinitions(e => e.Star(2).Star().Star(3))
.ColumnDefinitions(e => e.Absolute(100).Star());
``` 

In this example you define
- 3 rows - `Star(2)`, `Star()`, `Star(3)`
- 2 columns - `Absolute(100)`, `Star()`

### `BindableProperty` => `Setter`

```cs
Button.BackgroundColorProperty.Set(Colors.White),

Button.TextColorProperty.Set().OnLight(Colors.White).OnDark(AppColors.Primary),
```

implementation

```cs
public static class BindablePropertyExtension
{
    public static Setter Set(this BindableProperty property, object value) => new Setter { Property = property, Value = value };
    public static Setter Set(this BindableProperty property) => new Setter { Property = property };		
}

public static class SetterExtension
{
    public static Setter OnLight(this Setter setter, object value) { if (Application.Current?.RequestedTheme == AppTheme.Light) setter.Value = value; return setter; }
    public static Setter OnDark(this Setter setter, object value) { if (Application.Current?.RequestedTheme == AppTheme.Dark) setter.Value = value; return setter; }
    ...
}
```

### `IList` Add methods

Additional `Add()` methods for the `IList` interface

```cs
public partial class Style
{
    public void Add(Setter setter) {...}
    public void Add(Trigger trigger) {...}
    public void Add(DataTrigger trigger) {...}
    public void Add(VisualStateGroupList groupList) {...}
    public void Add(VisualStateGroup group) {...}
    public void Add(VisualState visualState) {...}
}
```

Usage example
```cs
new Style(typeof(Button))
{
    Button.TextColorProperty.Set().OnLight(Colors.White).OnDark(AppColors.Primary),
    Button.BackgroundColorProperty.Set().OnLight(AppColors.Primary).OnDark(Colors.White),
    Button.FontSizeProperty.Set(14).OnDesktop(20),
    Button.CornerRadiusProperty.Set(8).OniOS(15),
    new VisualState(VisualState.VisualElement.Normal)
    {
        Button.TextColorProperty.Set().OnLight(Colors.White).OnDark(AppColors.Primary),
        Button.BackgroundColorProperty.Set().OnLight(AppColors.Primary).OnDark(Colors.White),
    },
    new VisualState(VisualState.VisualElement.Disabled)
    {
        Button.TextColorProperty.Set().OnLight(AppColors.Gray950).OnDark(AppColors.Gray200),
        Button.BackgroundColorProperty.Set().OnLight(AppColors.Gray200).OnDark(AppColors.Gray600),
    },
},
```

### `[FluentInterface]` attribute

Fluent methods will be generated for user definied classes with the `[FluentInterface]` attribute

### `[BindableProperties]` attribute

Bindable properties and fluent methods will be generated for user definied classes with interfaces with the `[BindableProperties]` attribute

##### Additional atributes:

- `[DefaultValue]` to define default values
- `[PropertyCallbacks(propertyChanged, propertyChanging,validateValue, coerceValue, defaultValueCreator]` to define callback names

##### Example

```cs
[BindableProperties]
public interface IAngleViewModelProperties
{
    [PropertyCallbacks(propertyChanged: "OnAngleChanged")]
    public double RawAngle { get; set; }

    [PropertyCallbacks(coerceValue: "CoerceAngle")]
    public double Angle { get; set; }

    [DefaultValue(360.0)]
    public double MaximumAngle { get; set; }
}


[FluentInterface]
public partial class AngleViewModel : BindableObject, IAngleViewModelProperties
{
    static void OnAngleChanged(BindableObject bindable, object oldValue, object newValue)
    {
        var viewModel = bindable as AngleViewModel;
        viewModel.Angle = (double)newValue;
    }

    static object CoerceAngle(BindableObject bindable, object value)
    {
        var viewModel = bindable as AngleViewModel;
        double input = (double)value;

        if (input > viewModel.MaximumAngle)
            input = viewModel.MaximumAngle;

        return input;
    }
}
```

## Custom user view example

```cs
[BindableProperties]
public interface ICardViewProperties
{
    string CardTitle { get; set; }
    string CardDescription { get; set; }
    Color CardColor { get; set; }
    Color BorderColor { get; set; }
    Style DescriptionStyle { get; set; }
    View ContentView { get; set; }
    string ButtonTitle { get; set; }
}

[FluentInterface]
[ContentProperty(nameof(ContentView))]
public partial class CardView : ContentView, ICardViewProperties
{
    public event EventHandler Clicked;

    public CardView()
    {
        this.BindingContext = this;
        Content = new Border
        {
            new Grid
            {
                new VerticalStackLayout
                {
                    new Label(out var label1)
                        .Text(e => e.Path(nameof(CardTitle)))
                        .FontSize(29)
                        .TextColor(Colors.White),

                    new Label()
                        .Text(e => e.Path(nameof(CardDescription)))
                        .Style(e => e.Path(nameof(DescriptionStyle))),
                },

                new ContentView()
                    .Row(1)
                    .Content(e => e.Path(nameof(ContentView)))
                    .HorizontalOptions(LayoutOptions.Center)
                    .VerticalOptions(LayoutOptions.Center)
                    .SizeRequest(120,120),

                new Button()
                    .Row(2)
                    .Text(e => e.Path(nameof(ButtonTitle)))
                    .BackgroundColor(Colors.LightGray)
                    .TextColor(Colors.Black)
                    .OnClicked((sender, e) => Clicked(sender,e))
            }
            .RowDefinitions(e => e.Star(1.0).Star(2.0).Star(0.7))
            .RowSpacing(10)
        }
        .StrokeShape(new RoundRectangle().CornerRadius(10))
        .Stroke(e => e.Path(nameof(BorderColor)))
        .BackgroundColor(e => e.Path(nameof(CardColor)))
        .SizeRequest(200, 300)
        .Margin(50)
        .Padding(20);
    }
}

public partial class CardViewPage : ContentPage
{
    public CardViewPage()
    {
        this.Content = new VerticalStackLayout
        {
            new Slider(1, 100, 1, out var slider),

            new HorizontalStackLayout
            {
                new CardView(out var cardNo1)
                {
                    new Image("dotnet_bot.png").Aspect(Aspect.AspectFit)
                }
                .CardTitle(e => e.Path("Value").Source(slider).StringFormat("Value {0:F1}"))
                .ButtonTitle("Play")
                .CardDescription("Do you like it")
                .CardColor(Colors.DarkSlateGrey)
                .BorderColor(Colors.DarkGrey)
                .OnClicked(e =>
                {
                    cardNo1.CardDescription = "Let's play :)";
                }),

                new CardView(out var cardView)
                {
                    new VerticalStackLayout
                    {
                        new Label("This is a simple card view example"),
                        new Label("Second label")
                            .TextColor(Colors.Red)
                            .FontSize(18)
                    }					
                }
                .CardTitle("Title 2")
                .ButtonTitle("Stop")
                .CardColor(Colors.DarkSlateGrey)
                .BorderColor(Colors.DarkGrey)
                .DescriptionStyle(new Style(typeof(Label))
                {
                    Label.TextColorProperty.Set(Colors.Blue),
                    Label.FontSizeProperty.Set(20)
                })
            }
            .HorizontalOptions(LayoutOptions.Center)
        }
        .VerticalOptions(LayoutOptions.Center)
        .Padding(100);
    }
}
```

**Comment by mattleibow** [comment]
---
**Issue moved from dotnet/maui#12380**
- Please respond to @Ghostbird.

---

_From @Ghostbird on Tuesday, January 3, 2023 7:09:24 AM_

There is a [Fluent API for MAUI](https://github.com/CommunityToolkit/Maui.Markup). How did you managed to write one yourself without finding that one first? I think they'd welcome your contributions.

**Comment by mattleibow** [comment]
---
**Issue moved from dotnet/maui#12380**
- Please respond to @saint4eva.

---

_From @saint4eva on Tuesday, January 3, 2023 7:40:20 AM_

> There is a [Fluent API for MAUI](https://github.com/CommunityToolkit/Maui.Markup). How did you managed to write one yourself without finding that one first? I think they'd welcome your contributions.

His library feels more natural to me than C# for MarkUp. But both are heavily MVVM oriented. I personally love MVU and that is why I love [Comet (C# MVU)](https://github.com/dotnet/Comet). But I have been playing around with [Sharp.UI](https://github.com/idexus/Sharp.UI) library and I am liking it its simplicity - the same as Comet.

**Comment by mattleibow** [comment]
---
**Issue moved from dotnet/maui#12380**
- Please respond to @idexus.

---

_From @idexus on Tuesday, January 3, 2023 8:43:19 AM_

@Ghostbird I got to know this library, unfortunately it doesn't cover all properties in fluid methods because their approach is manual.

```cs
new Entry()
  .Keyboard(Keyboard.Numeric) // <<< compile error in Community Markup
  .FontSize(28)
```

I decided to create my lib using source generators. This allowed me to adopt in-line builders, e.g. to create bindings in more intuitive way.

```cs
new VerticalStackLayout
{
  new Slider(1, 100, 1, out var slider),
  new Label()
      .Text(e => e.Path("Value").StringFormat("Slider Value : {0}"))
      .FontSize(e => e.Path("Value"))
      .RotationX(e => e.Path("Value")),
}
```

Similarly, it is not possible in Community Markup to do something like this:

```cs
new ScrollView
{
    new Border
    {
        new Grid
        {
            new VerticalStackLayout
            {
                new Label("foo")
                    .FontSize(30),
                new Label("test")
            },

            new Button()
                .Row(1)
                .OnClicked(button =>
                {

                })
        }
        .RowDefinitions(e => e.Star(3).Star(1))
    }
    .StrokeShape(new RoundRectangle().CornerRadius(10))
    .SizeRequest(200, 300)
};
```

It can not be done by only extension methods.

**Comment by idexus** [comment]
@mattleibow  I created an issue but I see that it was redirected here. Personally, I think that to do it right, in an intuitive way, you need to create full support for using MAUI in the code, nowhere else but within the MAUI project. Many of these things can be done with source generators directly in the MAUI project as I have proposed. This cannot be done with extension methods and Community Markup alone. It's just patching holes.

**Reply by idexus** [reply]
@mattleibow It took me only 2 days to port my library to the maui project so I think it's doable. You can try Maui.Controls.Sample.Fluent in my fork [https://github.com/idexus/maui](https://github.com/idexus/maui).

Unfortunately, some things can't be done without getting inside the maui project. Another solution is subclassing, unfortunately some maui classes are sealed, so you can't do it without a wrapper. That's why I asked to unseal them.

**Reply by idexus** [reply]
You can judge it yourself, it's form your documentation

#### without

```cs
Grid grid = new Grid();

Label label = new Label { Text = "Code: " };
grid.Children.Add(label, 0, 1);

Entry entry = new Entry
{
    Placeholder = "Enter number", Keyboard = Keyboard.Numeric, BackgroundColor = Color.AliceBlue, TextColor = Color.Black, FontSize = 15,
    HeightRequest = 44, Margin = fieldMargin
};
grid.Children.Add(entry, 0, 2);
Grid.SetColumnSpan(entry, 2);
entry.SetBinding(Entry.TextProperty, new Binding("RegistrationCode"));

Content = grid;
```

#### Community Markup:

```cs
Content = new Grid { Children =
{
    new Label { Text = "Code:" }
               .Row (BodyRow.CodeHeader) .Column (BodyCol.Header),

    new Entry { Placeholder = "Enter number", Keyboard = Keyboard.Numeric, BackgroundColor = Color.AliceBlue, TextColor = Color.Black } .FontSize (15)
               .Row (BodyRow.CodeEntry) .ColumnSpan (All<BodyCol>()) .Margin (fieldMargin) .Height (44)
               .Bind (nameof(vm.RegistrationCode))
}}};
```

#### my approach in Sharp.UI:

```cs
Content = new Grid
{
    new Label("Code:").Row(1),
    new Entry("Enter number").Row(2).ColumnSpan(2)
        .Keyboard(Keyboard.Numeric)
        .BackgroundColor(Colors.AliceBlue)
        .TextColor(Colors.Black)
        .FontSize(15)
        .Margin(fieldMargin)
        .HeightRequest(44)
        .Text(e => e.Path("RegistrationCode"))
};
```
But as I said before, many of these things cannot be done without getting inside the maui project.

**Reply by saint4eva** [reply]
I think I support what @idexus is saying. The issue request primarily was to open up the sealed classes so he could generate more succinct fluent API without wrapping them. I would not know why you moved it here @mattleibow

**Reply by idexus** [reply]
Thank you for your support @saint4eva. 

I opened a [pull request](https://github.com/dotnet/maui/pull/12451) to open sealed classes that I use in my project. I hope they agree to it. And the fact, I also do not understand why the issue I created was moved here @mattleibow

**Comment by idexus** [comment]
@mattleibow 

> Moving this to the markup issue as we can discuss it there. This repo probably will not get this feature as it is more for the "core" feature and not the API variations.

I do not agree with you. Try to do something like this in the Community Markup without going into "core".

```cs
public class AlternateCollectionPage : ContentPage
{
    class AlternateRowModel
    {
        public int Id { get; set; }
        public object DataModel { get; set; }
        public Style LabelStyle => new Style<Label> { Label.TextColorProperty.Set(Id % 2 == 0 ? Colors.DarkSlateGray : Colors.LightGray) };
        public Style GridStyle => new Style<Grid> { Grid.BackgroundProperty.Set(Id % 2 == 0 ? Colors.LightGray : Colors.DarkSlateGray) };
    }

    public AlternateCollectionPage()
    {          
        var itemSource = DataModel.SimpleData
            .Select((e, i) => new AlternateRowModel { DataModel = e, Id = i });

        Content = new VStack
        {
            new CollectionView
            {
                () => new Grid()
                {
                    new Label()
                        .Text(e => e.Path("DataModel.Name"))
                        .Style(e => e.Path("LabelStyle"))

                }
                .Style(e => e.Path("GridStyle"))
            }
            .ItemsSource(itemSource)
        };
    }
}
```

**Comment by TheCodeTraveler** [comment]
Hi Everyone! There seems to be 2 main topics here:
- ~Create new Constructors for .NET MAUI controls~
  - Edit: This is not possible until [partial classes support external assemblies](https://github.com/dotnet/csharplang/issues/5497)
- Create Extension methods for every missing .NET MAUI Property

First, thank you so much for this Discussion! I'm _very_ excited to see this added to [`CommunityToolkit.Maui.Markup`](https://github.com/CommunityToolkit/Maui.Markup). If you haven't yet heard of the .NET MAUI Community Toolkits, it is a Community-driven, Microsoft-supported, library that we use as a staging ground for new features, contributed by the community, that will eventually be promoted into .NET MAUI.

For more information, here's a blog post I wrote when we released the .NET MAUI Community Toolkits: https://devblogs.microsoft.com/dotnet/introducing-the-net-maui-community-toolkit-preview/

Let's pave a path forward here for anyone from the community who would help with this effort!

~## Overloaded Constructors~

**Edit**: This is not possible until [partial classes support external assemblies](https://github.com/dotnet/csharplang/issues/5497)

Most [.NET MAUI Controls](https://github.com/dotnet/maui/tree/main/src/Controls/src/Core) are already using a `partial class`.

This means that we can create a second `partial class`, eg `partial class Button` that has a constructor accepting every Property as a default parameter.

~### Implementation~

**Edit**: This is not possible until [partial classes support external assemblies](https://github.com/dotnet/csharplang/issues/5497)

- Create a partial classes for each .NET MAUI Control
- Create a new Constructor that accepts every Property as a default parameter
- Assign the Properties from their appropriate parameters in the constructor 

~### Example~

**Edit**: This is not possible until [partial classes support external assemblies](https://github.com/dotnet/csharplang/issues/5497)

```cs
namespace Microsoft.Maui.Controls;

public partial class Button
{
  public Button(Thickness padding = default, 
                  LineBreakMode lineBreakMode = default,
                  Thickness margin = default
                  // continue adding a parameter for every Property of `Button`, including the properties inherited by `View`
                  )
  {
    this.Padding = padding;
    this.LineBreakMode = lineBreakMode;
    this.Margin = margin;
    // continuing assigning every parameter to its Property 
  }
}
```

~### Next Steps~

In the [`CommunityToolkit.Maui.Markup` repo](https://github.com/CommunityToolkit/Maui.Markup), create a [New Feature Proposal](https://github.com/CommunityToolkit/Maui.Markup/issues/new?assignees=&labels=new%2Cproposal&template=open-a-new-feature-proposal.yml&title=%5BProposal%5D+) for each .NET MAUI Control. We need an individual Proposal for each control so that we divide the work into smaller, more manageable PRs, and to ensure we don't forget any controls.

After each Proposal has been created, the Community Toolkit core maintainer team will review the Proposal and help refine the details. The team will then vote to Approve the implementation of each Proposal at [our monthly standup](https://www.youtube.com/watch?v=N2OZzLNSD_4).

After each Proposal has been approved, any community member is welcome to submit a PR implementing it.

## Extension Methods for .NET MAUI Properties

Believe it or not, this is already an existing goal of [`CommunityToolkit.Maui.Markup`](https://www.nuget.org/packages/CommunityToolkit.Maui.Markup/). It currently has extension methods for _most_ Properties, but it is still missing extension methods for many properties.

### Implementation

- Create `public static class` for each .NET MAUI class, eg `public static class CollectionViewExtensions`
  - See [`ViewExtensions`](https://github.com/CommunityToolkit/Maui.Markup/blob/main/src/CommunityToolkit.Maui.Markup/ViewExtensions.cs) as an example
- Add a Fluent Extension method that accepts a Generic Type based on the base type, and returns the same Generic Type

### Example
Here's an existing example from `CommunityToolkit.Maui.Markup` for  [`ViewExtensions.CenterHorizontal()`](https://github.com/CommunityToolkit/Maui.Markup/blob/cf4b8e3232386bb13b5d65f158e62038fb07c750/src/CommunityToolkit.Maui.Markup/ViewExtensions.cs#LL26C3-L30C4):
  ```cs
    public static TView CenterHorizontal<TView>(this TView view) where TView : View
    {
	    view.HorizontalOptions = LayoutOptions.Center;
	    return view;
    }
  ```
### Next Steps

In the [`CommunityToolkit.Maui.Markup` repo](https://github.com/CommunityToolkit/Maui.Markup), create a [New Feature Proposal](https://github.com/CommunityToolkit/Maui.Markup/issues/new?assignees=&labels=new%2Cproposal&template=open-a-new-feature-proposal.yml&title=%5BProposal%5D+) for each .NET MAUI Control. We need an individual Proposal for each control so that we divide the work into smaller, more manageable PRs, and to ensure we don't forget any controls.

After each Proposal has been created, the Community Toolkit core maintainer team will review the Proposal and help refine the details. The team will then vote to Approve the implementation of each Proposal at [our monthly standup](https://www.youtube.com/watch?v=N2OZzLNSD_4).

After each Proposal has been approved, any community member is welcome to submit a PR implementing it.

**Reply by pictos** [reply]
@brminnick are you suggesting creating the `partial classes` with new ctors on our lib?

**Reply by TheCodeTraveler** [reply]
Yup! Will it work if our `partial class` is in a different assembly than the .NET MAUI `partial class`?

**Reply by pictos** [reply]
No. we can't do `partial class` on a "foreign" assembly. I tried that while a go to try to simplify control creations... But that could work in the future if the C# team implement [this feature](https://github.com/dotnet/csharplang/issues/5497). Not sure if using `ctors` or another fancy API idea

**Reply by idexus** [reply]
@brminnick @mattleibow @pictos @davidortinau @PureWeen

That's why I started a discussion, you can't create a `partial class` declaration from another assembly.
#

I didn't want to use XAML, so at the beginning of my journey with MAUI I started using your Community Markup. Unfortunately, I noticed a lot of holes, and that, I don't want to offend anyone here, but things were done in many places very unintuitively, the design lacked coherence. The method-by-method approach of adding functionality to the library scared me. Not only does this result in very slow functionality growth, it actually causes a loss of coherence when these things could be generated.

That's why I decided to create my own library and rely on source generators in it. It took me about 3 weeks to get a usable version. Unfortunately, to add some functionalities, I couldn't do it in a simple way, because you can't add other functionalities to classes besides extension methods. This caused me to subclass. Unfortunately, some classes are sealed, so I had to create a wrapper for them. My library works, but I think it can be done in a simpler and more correct way by implementing some of the functionality in the MAUI core.

I think you should create a new approach, not destroy what is already there, because too much code depends on it. You need to build a solid architecture for the new API, from the ground up. Maybe you will have to abandon the current idea of Community Markup in the future. Many of the things I wrote about earlier should be done with source generators. It's pointless to create tons of extension methods manually when they can be generated. There is a need to meet, to build a solid foundation, and then you can follow them. You can't rely on the community for everything and patch hole by hole. Without solid foundations, good architecture, the project, in my opinion, will be exposed to failure.

The concept of MAUI itself is good, but you need to give people a good tool to make them want to use it in their projects. XAML, no offense, sucks. He ties a man's hands in many places. I think the Fluent API could become a first-class way to develop applications in MAUI, but as I said, it can't be done in a community-only way, without good architecture and of course without full support in the MAUI core.

Otherwise, I'm sorry to say it here, but in the near future, others will be light years ahead of you.

**Reply by TheCodeTraveler** [reply]
> others will be light years ahead of you

Thanks @idexus. I think you intended to write this to the .NET MAUI team. We are the Community Toolkit team and do not work for Microsoft and cannot influence the backlog of the .NET MAUI team. 

Here's the link to the .NET MAUI repo where you can raise your concerns: https://github.com/dotnet/maui

**Reply by idexus** [reply]
@brminnick I understand to a large extent that your hands are tied. I am very sad about this. Originally my post was directed there, but it was redirected to you. I opened a [pull request](https://github.com/dotnet/maui/pull/12451) for unsealing some classes, this would enable a lot. We'll see if they agree to this change.

**Reply by saint4eva** [reply]
@brminnick Thanks for your comments and suggestion. However, accepting a PR from @idexus means there will be conspicuous change in the syntax of C# for Markup. Looking at the above examples from the two libraries, you will notice they are very different in styles. So, my question is how do you reconcile the two different styles. And also, it would be important for the Maui team to unseal some classes to make it easier for more people to come up with a C# based solutions to writing UI (choices are welcome - healthy community). Unless unsealing the classes affects the performance of the Maui Core?
